WooCommerce/Classes/ViewRelated/Dashboard/Settings/Help/ApplicationLogViewModel.swift

```diff
@@ -0,0 +1,130 @@
+            self.date = nil
+            self.text = text
+            return
+        }
+
+        self.date = date
+        self.text = String(text[logRange])
+    }
+}
+
+final class ApplicationLogViewModel: ObservableObject {
+    private let logText: String
+    let lines: [ApplicationLogLine]
+
+    let lastLineID: UUID
+
+    var present: ((UIViewController) -> Void)? = nil
+
+    @Published var lastCellIsVisible = false
+
+    @Published var buttonVisible = true
+
+    private var cancellableSet: Set<AnyCancellable> = []
+
+    init(logText: String) {
+        self.logText = logText
+        lines = logText
+            .split(separator: "\n")
+            .map(String.init)
+            .map(ApplicationLogLine.init(text:))
+
+        // The `ScrollViewProxy.scrollTo` method doesn't seem to handle an optional line well
+        // Instead of `nil`, we can represent a missing value with a random UUID
+        lastLineID = lines.last?.id ?? UUID()
+
+        guard lines.isNotEmpty else {
+            // If the file is empty, there will be no last cell,
+            // and so the last cell won't be visible, but it still doesn't make
+            // sense to show a button to scroll down
+            buttonVisible = false
+            return
+        }
+
+        $lastCellIsVisible
+            .map(!)
+            .debounce(for: .milliseconds(100), scheduler: DispatchQueue.main)
+            .assign(to: \.buttonVisible, on: self)
+            .store(in: &cancellableSet)
+    }
+
+    func isLastLine(_ line: ApplicationLogLine) -> Bool {
+        line.id == lastLineID
+    }
+
+    func showShareActivity() {
+        let activityVC = UIActivityViewController(activityItems: [logText], applicationActivities: nil)
+        activityVC.modalPresentationStyle = .popover
+        activityVC.excludedActivityTypes = assembleExcludedSupportTypes()
+        present?(activityVC)
+    }
+
+    /// Specifies the activity types that should be excluded.
+    /// - returns: all unsupported types
+    /// Preserves support for `.copyToPasteboard`, `.mail`, and `.airDrop`
+    ///
+    func assembleExcludedSupportTypes() -> [UIActivity.ActivityType] {
+        let activityTypes = NSMutableSet(array: SharingHelper.allActivityTypes())
+
+        /*
+         * Don't use Set(arrayLiteral:) here, because it will convert the enums to the raw string value,
+         * which would be wrong, because the allActivityTypes listed above are stored as enum types.
+         */
+        let supportedTypes = NSSet(objects: UIActivity.ActivityType.copyToPasteboard,
+                                            UIActivity.ActivityType.mail,
+                                            UIActivity.ActivityType.airDrop)
+
+        // Now you can downcast to Set, because the type was preserved on init of the NSSet.
+        activityTypes.minus(supportedTypes as! Set<AnyHashable>)
+
+        return activityTypes.allObjects as! [UIActivity.ActivityType]
```

Comment:
I know this is legacy code but I still have nightmares when I see `!` in production code!

Reply:
I didn't try to update this code at first, but this was probably written when `Set` was much less powerful. I've been able to rewrite it now without all the weird casting 🎉